(this PR is just for my personal  self-learning) [PaddlePaddle Hackathon 2 No.22]  add paddle.index_add  to Paddle

OccupyMars2030 opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
New features
### PR changes
<!-- One of [ OPs | APIs | Docs | Others ] -->
APIs
### Describe
<!-- Describe what this PR does -->
[【PaddlePaddle Hackathon 第二期】任务总览](https://github.com/PaddlePaddle/Paddle/issues/40234#)
task 22:  add paddle.index_add  to Paddle
https://github.com/PaddlePaddle/Paddle/issues/40311

RFC：  https://github.com/PaddlePaddle/community/pull/127

This PR is  gotten by  modifying  this PR: https://github.com/PaddlePaddle/Paddle/pull/42454
(with least  modification)

I stop  working on  this PR   at   2022/5/17  18:00 .  I  will  continue  to  work  on  this  PR  after  2022/6/10  when  I  have time.

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

dingjiaweiww commented:
你好~按照活动流程，请先补充该题的RFC~

OccupyMars2030 commented:
如果需要的话, 我可以立刻补充该题的RFC

TCChenlong commented:
> 如果需要的话, 我可以立刻补充该题的RFC

这个是需要的哈 请在 https://github.com/PaddlePaddle/community/tree/master/rfcs/APIs 这个目录下，按规范提交PFC~
规范请参考：https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/dev_guides/api_contributing_guides/api_contributing_guides_cn.html

OccupyMars2030 commented:
收到,  我今天之内完成

OccupyMars2030 commented:
done.

https://github.com/PaddlePaddle/community/pull/127

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Update __init__.py

OccupyMars2030 committed:
Merge branch 'develop' into hackathon-2nd-task22-add-index_add

OccupyMars2030 committed:
Merge branch 'PaddlePaddle:develop' into hackathon-2nd-task22-add-index_add

iclementine commented on paddle/fluid/operators/index_add_op.cc:
```diff
@@ -0,0 +1,115 @@
+        .SetDefault(0);
+    AddAttr<float>("added_value",
+                   "(float, default 0.0f) The value to add.")
+        .SetDefault(0.0f);
```
不建议添加默认值，因为没有合理的默认值

iclementine commented on paddle/fluid/operators/index_add_op.cc:
```diff
@@ -0,0 +1,115 @@
+    AddAttr<int>("axis",
+                 "(int, default 0), "
+                 "the dimension in which we index.")
+        .SetDefault(0);
```
不建议添加默认值，因为没有合理的默认值

此外，axis 希望也能支持 Tensor 输入，亦即也作为一个 Input 而不只是 Attr. 具体实现方法可以参考 `paddle/fluid/operators/scale_op.cc`

iclementine commented on paddle/fluid/operators/index_add_op.cc:
```diff
@@ -0,0 +1,115 @@
+             "(Tensor, default input Tensor<float>), "
+             "the input feature data of IndexAddOp, dtype should be"
+             "int32, int64, float16, float32, float64.");
+    AddInput("Index",
```
Index 希望也能支持 python 的 list 或者 tuple of ints.
这样写是不能支持传入 list or tuple of ints, 这和设计中的描述也不一致。

iclementine commented on python/paddle/tensor/manipulation.py:
```diff
@@ -4238,6 +4238,112 @@ def put_along_axis_(arr, indices, values, axis, reduce='assign'):
+            "The numel of added_value must be one when it is a tensor.")
+
+
+def index_add(x, index, axis=0, added_value=0.0):
```
不建议添加默认值。

iclementine commented on paddle/phi/infermeta/binary.h:
```diff
@@ -184,6 +184,17 @@ void HuberLossInferMeta(const MetaTensor& input_meta,
+void IndexAddInferMeta(const MetaTensor& x,
+                       const MetaTensor& index,
+                       int axis,
+                       float added_value,
```
如前面所属，axis, index, add_value 都支持 Tensor 或者 非 Tensor 的话，这些也都需要相应修改。

iclementine commented on paddle/phi/kernels/cpu/index_add_kernel.cc:
```diff
@@ -0,0 +1,66 @@
+                   phi::dtype::float16,
+                   double,
+                   int,
+                   int64_t) {}
```
希望也支持 bool, bfloat16 和 complex 数据类型。
GPU 环境下也是，以及反向也是如此。

iclementine commented on paddle/phi/kernels/gpu/index_add_grad_kernel.cu:
```diff
@@ -0,0 +1,42 @@
+#include "paddle/phi/kernels/copy_kernel.h"
+#include "paddle/phi/kernels/index_add_grad_kernel.h"
+
+DECLARE_bool(cudnn_deterministic);
```
如果没有用到这个，不需要定义。

iclementine reviewed (commented):
主要的问题是 

1. axis, index，add_value 都希望能支持 Tensor 类型和非 Tensor 类型的数据，axis 需要支持 int 和 0D 整形 Tensor, index 需要支持 list or tuple of ints 或者 1D 整形 Tensor, add_value 需要支持 python scalalr 类型以及多维 Tensor, 只要其形状可以 broadcast 到切片的形状；
2. add_value 是 Tensor 的情况，需要支持梯度回传到 add_value;
3. 补齐数据类型支持；

会需要比较大的修改

OccupyMars2030 commented:
感谢Paddle专家详细的回复，我尽快修改完后重新提交

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

SmirnovKol commented:
1.  关于axis:  我准备在python端用
 `axis = int(axis)`  ， 然后再调用op算子（op算子中的axis 就只是作为int  attribute）
2.  关于index:  我准备在C++端将index作为IntArray  attribute,
在python端调用op算子时，index可以作为list or tuple of ints 或者 1-D tensor of  dtype int 传入 op算子 （这样做是否正确，我还在研究）
3.  关于add_value: 目前计划是在OpMaker中：
`AddInput("AddValueTensor", "....")  (dispensable)
AddAttr<phi::Scalar>("add_value", "...")
`
AddValueTensor 的优先级要高于 add_value,  我还在研究怎么写（也许还可以参考dropout中seed的实现）

4.  关于index:  目前我会进行算前检查，要求里面的元素必须在 [0,  x.shape[axis]) 范围内， **且必须无重复**，主要是按照我的理解，在CUDA计算时，对于input 的所有要修改的元素，一个线程负责一个元素，多线程并行互不干扰，如果index中有重复的元素，我担心可能会有线程冲突，我不清楚如何解决

SmirnovKol commented:
目前基本就是以上的问题，我在研究别的类似算子是怎么写的，另外感觉_C_ops.index_add()这种调用似乎与pybind有关系，所以也在尝试了解pybind。今天是周五，周日是修改提交的截至日，所以希望paddle专家能在今天再提供一些指导意见或者可以参考的算子，谢谢

SmirnovKol commented:
Boost.Python的引用头文件是<boost/python.hpp>，命名空间是boost::python.  BOOST.Python，这是BOOST开发组专门为C++/Python直接的互操作而开发的类库，也是Boost中唯一一个解释语言支持模块。Boost.Python除了提供Python C API的C++兼容封装，也使得整个扩展/嵌入过程更加方便。

SmirnovKol commented:
熟悉以下技术和概念:
Boost
Boost.Python
PyObject

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Merge branch 'PaddlePaddle:develop' into hackathon-2nd-task22-add-index_add

OccupyMars2030 committed:
Update unary.cc

OccupyMars2030 committed:
Update index_add_op.cc

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 commented:
CI报错，似乎之前其他人merge到develop分支的pr有问题

OccupyMars2030 commented:
![Screenshot from 2022-05-16 05-58-47](https://user-images.githubusercontent.com/31559413/168496065-9f53009e-be39-412e-bcf3-628546dfd3a8.png)

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 commented:
还是有其他算子的单侧报错了
![Screenshot from 2022-05-16 11-38-43](https://user-images.githubusercontent.com/31559413/168515875-47002cfb-7bfe-4115-baef-14a00a6cfb2e.png)

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Add files via upload

add  "data = np.copy(data)" in the first line  in the function  "def index_add_np(data, axis, index, add_value):"  in  unit test file

OccupyMars2030 commented:
Caution:  add  "x = np.copy(x)"  at the first line of the function that use  numpy to simulate the calculation of  paddle op  in  unit  test file.   Because  if  you don't  use the  copy  of  input  at  first , then  the calculation  in the  function  may  change the  elements of  numpy.ndarray   in place,   so  after  calling  this  function,  the  input   numpy.ndarray  is  modified  to be equal  to  the output  numpy.ndarray

TODO:  https://github.com/PaddlePaddle/Paddle/pull/42454
in the above PR,   there seems to  be  no  such  'copy'  thing,  how it  passes  the  unit  test ???
remain  to  be  studied...

OccupyMars2030 committed:
Add files via upload

OccupyMars2030 committed:
Merge branch 'PaddlePaddle:develop' into hackathon-2nd-task22-add-index_add

OccupyMars2030 commented:
![Screenshot from 2022-05-16 18-37-30](https://user-images.githubusercontent.com/31559413/168575361-342b83da-888a-44f5-aea3-7d71448738ba.png)
再次触发CI后，发现报错信息完全与我的代码无关，是其他算子的单测有问题
解决方法： Merge  [PaddlePaddle:develop](https://github.com/PaddlePaddle/Paddle)  到我的分支，重新触发CI

OccupyMars2030 commented:
![Screenshot from 2022-05-16 22-44-08](https://user-images.githubusercontent.com/31559413/168619689-829952f2-5281-4940-8e0d-06c6676c26d8.png)
已经Merge [PaddlePaddle:develop](https://github.com/PaddlePaddle/Paddle) 到我的分支，重新触发CI，发现报错信息仍然完全与我的代码无关，是其他算子的单测有问题

OccupyMars2030 committed:
Merge branch 'PaddlePaddle:develop' into hackathon-2nd-task22-add-index_add

OccupyMars2030 committed:
just trigger CI again

OccupyMars2030 commented:
![Screenshot from 2022-05-17 12-23-46](https://user-images.githubusercontent.com/31559413/168728550-219e36da-da16-4d54-a2c5-0011f45103ce.png)

重新触发CI，发现报错信息仍然完全与我的代码无关，是其他算子的单测有问题
解决办法：提交其他PR，看是否是所有的PR都有这个问题

OccupyMars2030 commented:
https://github.com/PaddlePaddle/Paddle/pull/42804
看看这个PR是否会报同样的错误,  显示没有报同样的错误，说明是你修改了与任务无关的代码

OccupyMars2030 committed:
add line 439 and line 440 EinsumInferMeta in  paddle/phi/infermeta/unary.cc (when did I delete the two lines???)

OccupyMars2030 commented:
![Screenshot from 2022-05-17 13-58-34](https://user-images.githubusercontent.com/31559413/168739509-735507b7-5ea2-4f98-b713-7fa6de754602.png)
你确实修改了与任务无关的代码

OccupyMars2030 committed:
restore code about heaviside API

OccupyMars2030 committed:
restore code about heaviside API

OccupyMars2030 committed:
uncomment other unit test cases

OccupyMars2030 committed:
leave only one unit test case

OccupyMars2030 commented:
![Screenshot from 2022-05-17 17-51-23](https://user-images.githubusercontent.com/31559413/168783796-256b5a2c-68e2-4fb2-8579-572d454ec53f.png)
网上搜索发现：该报错似乎都与PaddlePaddle 训练或者infer相关，似乎是CUDA, cuDNN 版本问题？？？

单测文件中只留下第一个test  case,  其他的全部注释掉，重新提交，又可以了， 没有上面的"Segmentation  fault"报错问题了

OccupyMars2030 commented:
感谢Paddle 专家的指导意见，我将暂停本PR的开发工作，在以后有时间时再继续开发
